File: backend/app/libraries/budget.py
"""Server-side spending-limit enforcement (mirrors gateway/budget.ts). All
amounts are in micro-units (6-decimal USDC)."""


def enforce_budget(requested_micro: float, ctx: dict) -> dict:
    remaining_per_investigation = ctx["maxPerInvestigation"] * 1_000_000 - ctx["investigationSpent"]
    remaining_session = ctx["sessionBudget"] * 1_000_000 - ctx["sessionSpent"]
    remaining_total = ctx["totalBudget"] * 1_000_000 - ctx["totalSpent"]

    if requested_micro <= 0:
        return {"allowed": False, "reason": "per_evidence",
                "amountMicro": requested_micro, "limitMicro": round(ctx["maxPerEvidenceCheck"] * 1_000_000),
                "remainingMicro": 0}

    if requested_micro > ctx["maxPerEvidenceCheck"] * 1_000_000:
        return {"allowed": False, "reason": "per_evidence",
                "amountMicro": requested_micro, "limitMicro": round(ctx["maxPerEvidenceCheck"] * 1_000_000),
                "remainingMicro": 0}
    if requested_micro > remaining_per_investigation:
        return {"allowed": False, "reason": "per_investigation",
                "amountMicro": requested_micro, "limitMicro": max(0, round(remaining_per_investigation)),
                "remainingMicro": max(0, round(remaining_per_investigation))}
    if requested_micro > remaining_session:
        return {"allowed": False, "reason": "session_budget",
                "amountMicro": requested_micro, "limitMicro": max(0, round(remaining_session)),
                "remainingMicro": max(0, round(remaining_session))}
    if requested_micro > remaining_total:
        return {"allowed": False, "reason": "total_budget",
                "amountMicro": requested_micro, "limitMicro": max(0, round(remaining_total)),
                "remainingMicro": max(0, round(remaining_total))}
    return {"allowed": True, "amountMicro": requested_micro,
            "limitMicro": round(ctx["maxPerEvidenceCheck"] * 1_000_000),
            "remainingMicro": max(0, round(remaining_per_investigation))}


def budget_block_reason(reason: str) -> str:
    return {
        "per_evidence": "Cost exceeds per-evidence-check limit",
        "per_investigation": "Exceeds per-investigation budget",
        "session_budget": "Exceeds session budget",
        "total_budget": "Exceeds total budget",
    }.get(reason, "Blocked by spending limit")

File: src/lib/gateway/budget.ts
import type { BudgetContext, BudgetDecision } from "../types";

/**
 * Server-side spending-limit enforcement for the Evidence Acquisition Gateway.
 *
 * All amounts are in micro-units (6 decimal USDC). The frontend can show
 * estimated totals, but the authoritative gate is here — the backend blocks
 * any acquisition whose cost exceeds the user's configured limits.
 */
export function enforceBudget(requestedMicro: number, ctx: BudgetContext): BudgetDecision {
  const remainingPerInvestigation =
    ctx.maxPerInvestigation * 1e6 - ctx.investigationSpent;
  const remainingSession = ctx.sessionBudget * 1e6 - ctx.sessionSpent;
  const remainingTotal = ctx.totalBudget * 1e6 - ctx.totalSpent;

  if (requestedMicro <= 0) {
    return {
      allowed: false,
      reason: "per_evidence",
      amountMicro: requestedMicro,
      limitMicro: Math.round(ctx.maxPerEvidenceCheck * 1e6),
      remainingMicro: 0,
    };
  }

  if (requestedMicro > ctx.maxPerEvidenceCheck * 1e6) {
    return {
      allowed: false,
      reason: "per_evidence",
      amountMicro: requestedMicro,
      limitMicro: Math.round(ctx.maxPerEvidenceCheck * 1e6),
      remainingMicro: 0,
    };
  }
  if (requestedMicro > remainingPerInvestigation) {
    return {
      allowed: false,
      reason: "per_investigation",
      amountMicro: requestedMicro,
      limitMicro: Math.max(0, Math.round(remainingPerInvestigation)),
      remainingMicro: Math.max(0, Math.round(remainingPerInvestigation)),
    };
  }
  if (requestedMicro > remainingSession) {
    return {
      allowed: false,
      reason: "session_budget",
      amountMicro: requestedMicro,
      limitMicro: Math.max(0, Math.round(remainingSession)),
      remainingMicro: Math.max(0, Math.round(remainingSession)),
    };
  }
  if (requestedMicro > remainingTotal) {
    return {
      allowed: false,
      reason: "total_budget",
      amountMicro: requestedMicro,
      limitMicro: Math.max(0, Math.round(remainingTotal)),
      remainingMicro: Math.max(0, Math.round(remainingTotal)),
    };
  }
  return {
    allowed: true,
    amountMicro: requestedMicro,
    limitMicro: Math.round(ctx.maxPerEvidenceCheck * 1e6),
    remainingMicro: Math.max(0, Math.round(remainingPerInvestigation)),
  };
}

/** Build a BudgetContext from a user + an investigation. */
export function buildBudgetContext(input: {
  prefs: { maxPerEvidenceCheck: number; maxPerInvestigation: number; sessionBudget: number; totalBudget: number };
  investigationSpent: number;
  sessionSpent: number;
  totalSpent: number;
}): BudgetContext {
  return {
    maxPerEvidenceCheck: input.prefs.maxPerEvidenceCheck,
    maxPerInvestigation: input.prefs.maxPerInvestigation,
    sessionBudget: input.prefs.sessionBudget,
    totalBudget: input.prefs.totalBudget,
    investigationSpent: input.investigationSpent,
    sessionSpent: input.sessionSpent,
    totalSpent: input.totalSpent,
  };
}
